Cg/lab3

pv6 opened the pull request:
Цифры 1, 2 - яркость источников (на 3 источника написал слишком много команд на один шейдер)
3 - обычный рендеринг
4 - Normal Distribution function
5 - Geometry function
6 - Fresnel function

pv6 committed:
Change geometry to a sphere

pv6 committed:
Add PBR implementation

pv6 committed:
Add grid of spheres with different materials

pv6 committed:
Add rendering modes for different functions

'3' - regular rendering
'4' - Norm Distribution function
'5' - Geometry function
'6' - Fresnel function

pv6 committed:
Add environment sphere

anthark commented on anim/Content/FresnelPixelShader.hlsl:
```diff
@@ -0,0 +1,9 @@
+
+float4 main(PixelShaderInput input) : SV_TARGET
+{
+    return float4(fresnel(
```
Здесь может получится (и получается =) ) странный результат, что эффект Френеля дает яркое пятно на поверхности, когда источник света светит на поверхность сзади. В полном шейдере это будет компенсировано значением dot(n,wi), но лучше все такие значения все-таки всегда клампить, поскольку иначе они могут неожиданно дать странные эффекты в совершенно другом месте, и придется долго и мучительно искать, что не так.

pv6 replied:
Исправлено

anthark commented on anim/Content/PBRInclude.cginc:
```diff
@@ -0,0 +1,108 @@
+
+float3 h(float3 wi, float3 wo)
+{
+    return normalize((wi + wo) / 2);
```
Здесь можно на 2 не делить - все равно вектор будет нормализован. Скорее всего, шейдерный компилятор и сам это деление уберет, но без него исходный код шейдера будет выглядеть более профессионально

pv6 replied:
Исправлено

anthark commented on anim/Content/PBRPixelShader.hlsl:
```diff
@@ -0,0 +1,14 @@
+    float3 wo = toCamera(input.worldPos);
+
+    float3 finalColor =
+        albedo * 0.56f +
```
Это аналог ambient, если я правильно понял?

pv6 replied:
Да, простенький ambient lighting, чтобы естественнее смотрелось

anthark commented on anim/Content/Sample3DSceneRenderer.cpp:
```diff
@@ -268,84 +292,197 @@ void Sample3DSceneRenderer::CreateDeviceDependentResources()
+    );
+
+    static const int
+        numLatitudeLines = 128,
```
Единственная серьезная проблема: если параметры поставить 16 и 16, например, пятно света будет выглядеть... неправильно. Какую длину имеет вектор, полученный линейной интерполяцией двух единичных векторов?

pv6 replied:
Не подумал об этом, добавил нормализацию нормали

anthark reviewed (changes requested):
Почти идеально. 

Есть одна серьезная проблема, которую я попрошу вас поправить, остальное - просто советы.

pv6 committed:
Fix problems

- normalize interpolated normals
- add clamping for faces facing away from light source in Fresnel function
- remove redundant division in h vector
